compiler/src/wasm_intrinsics.h

```diff
@@ -9,110 +9,190 @@
+
     WID(NULL, WI_IF_START, 0x40);
     WID(NULL, WI_LOOP_START, 0x40);
-    
+
     WIL(NULL, WI_LOCAL_GET, count_local);
     WID(NULL, WI_I32_CONST, 1);
     WI(NULL, WI_I32_SUB);
     WIL(NULL, WI_LOCAL_SET, count_local);
-    
+
     WIL(NULL, WI_LOCAL_GET, dest_local);
     WIL(NULL, WI_LOCAL_GET, count_local);
     WI(NULL, WI_PTR_ADD);
-    
+
     WIL(NULL, WI_LOCAL_GET, source_local);
     WIL(NULL, WI_LOCAL_GET, count_local);
     WI(NULL, WI_PTR_ADD);
-    
+
     WID(NULL, WI_I32_LOAD_8_U, ((WasmInstructionData) { 0, 0 }));
     WID(NULL, WI_I32_STORE_8, ((WasmInstructionData) { 0, 0 }));
-    
+
     WIL(NULL, WI_LOCAL_GET, count_local);
     WID(NULL, WI_I32_CONST, 0);
     WI(NULL, WI_I32_GT_S);
     WID(NULL, WI_COND_JUMP, 0x00);
-    
+
     WI(NULL, WI_LOOP_END);
     WI(NULL, WI_IF_END);
-    
+
     local_raw_free(mod->local_alloc, WASM_TYPE_INT32);
     local_raw_free(mod->local_alloc, WASM_TYPE_PTR);
     local_raw_free(mod->local_alloc, WASM_TYPE_PTR);
-    
+
     *pcode = code;
 }
 
 EMIT_FUNC_NO_ARGS(intrinsic_memory_fill) {
     bh_arr(WasmInstruction) code = *pcode;
-    
+
     // The stack should look like this:
     //     <count>
     //     <byte>
     //     <dest>
-    
+
     u64 count_local = local_raw_allocate(mod->local_alloc, WASM_TYPE_INT32);
     u64 byte_local  = local_raw_allocate(mod->local_alloc, WASM_TYPE_INT32);
     u64 dest_local  = local_raw_allocate(mod->local_alloc, WASM_TYPE_PTR);
-    
+
     WIL(NULL, WI_LOCAL_SET, count_local);
     WIL(NULL, WI_LOCAL_SET, byte_local);
     WIL(NULL, WI_LOCAL_SET, dest_local);
-    
+
     // count is greater than 0
     WIL(NULL, WI_LOCAL_GET, count_local);
     WID(NULL, WI_I32_CONST, 0);
     WI(NULL, WI_I32_GT_S);
-    
+
     WID(NULL, WI_IF_START, 0x40);
     WID(NULL, WI_LOOP_START, 0x40);
-    
+
     WIL(NULL, WI_LOCAL_GET, count_local);
     WID(NULL, WI_I32_CONST, 1);
     WI(NULL, WI_I32_SUB);
     WIL(NULL, WI_LOCAL_SET, count_local);
-    
+
     WIL(NULL, WI_LOCAL_GET, dest_local);
     WIL(NULL, WI_LOCAL_GET, count_local);
     WI(NULL, WI_PTR_ADD);
-    
+
     WIL(NULL, WI_LOCAL_GET, byte_local);
     WID(NULL, WI_I32_STORE_8, ((WasmInstructionData) { 0, 0 }));
-    
+
     WIL(NULL, WI_LOCAL_GET, count_local);
     WID(NULL, WI_I32_CONST, 0);
     WI(NULL, WI_I32_GT_S);
     WID(NULL, WI_COND_JUMP, 0x00);
-    
+
     WI(NULL, WI_LOOP_END);
     WI(NULL, WI_IF_END);
-    
+
     local_raw_free(mod->local_alloc, WASM_TYPE_INT32);
     local_raw_free(mod->local_alloc, WASM_TYPE_INT32);
     local_raw_free(mod->local_alloc, WASM_TYPE_PTR);
-    
+
     *pcode = code;
 }
 
+// IMPROVE: This is a stripped down/unoptimized version of memcmp that only checks equality
+// between two blocks of memory (and doesn't return -1/1). Returns 1 if the blocks of memory are equal,
+// and 0 otherwise.
+// NOTE: This implementation is unsafe and assumes both blocks of memory are at least size_in_bytes long.
+EMIT_FUNC_NO_ARGS(intrinsic_memory_equal) {
```

Comment:
New intrinsic implementation is here. Commenting so you can skip past all of the whitespace changes

Reply:
I think this looks good. There might be a way to shrink the number of instructions, but for now this will work well!

Reply:
Definitely! May need to change in the future anyway if a `memory.compare` instruction is added to the spec